File: middleware/types_test.go
package middleware

import (
	"testing"
	"time"
)

func TestDecision_IsAllowed(t *testing.T) {
	tests := []struct {
		name     string
		action   Action
		expected bool
	}{
		{"allow action returns true", ActionAllow, true},
		{"deny action returns false", ActionDeny, false},
		{"filter action returns false", ActionFilter, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			d := Decision{Action: tt.action}
			if got := d.IsAllowed(); got != tt.expected {
				t.Errorf("IsAllowed() = %v, want %v", got, tt.expected)
			}
		})
	}
}

func TestDecision_IsDenied(t *testing.T) {
	tests := []struct {
		name     string
		action   Action
		expected bool
	}{
		{"allow action returns false", ActionAllow, false},
		{"deny action returns true", ActionDeny, true},
		{"filter action returns false", ActionFilter, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			d := Decision{Action: tt.action}
			if got := d.IsDenied(); got != tt.expected {
				t.Errorf("IsDenied() = %v, want %v", got, tt.expected)
			}
		})
	}
}

func TestDecision_IsFiltered(t *testing.T) {
	tests := []struct {
		name     string
		action   Action
		expected bool
	}{
		{"allow action returns false", ActionAllow, false},
		{"deny action returns false", ActionDeny, false},
		{"filter action returns true", ActionFilter, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			d := Decision{Action: tt.action}
			if got := d.IsFiltered(); got != tt.expected {
				t.Errorf("IsFiltered() = %v, want %v", got, tt.expected)
			}
		})
	}
}

func TestCallerContext(t *testing.T) {
	ctx := CallerContext{
		Trusted:   true,
		Role:      "admin",
		UserID:    "user123",
		SessionID: "session456",
		Source:    "cli",
		Metadata:  map[string]interface{}{"key": "value"},
	}

	if ctx.Trusted != true {
		t.Error("Trusted should be true")
	}
	if ctx.Role != "admin" {
		t.Error("Role should be admin")
	}
}

func TestToolResult(t *testing.T) {
	result := ToolResult{
		Content:  "test content",
		Error:    nil,
		IsStream: false,
		Metadata: map[string]interface{}{"foo": "bar"},
	}

	if result.Content != "test content" {
		t.Error("Content mismatch")
	}
	if result.Error != nil {
		t.Error("Error should be nil")
	}
}

func TestToolDefinition(t *testing.T) {
	def := ToolDefinition{
		Name:        "test_tool",
		Description: "A test tool",
		InputSchema: map[string]interface{}{"type": "object"},
	}

	if def.Name != "test_tool" {
		t.Error("Name mismatch")
	}
}

func TestDecision_Timestamp(t *testing.T) {
	now := time.Now()
	d := Decision{
		Timestamp: now,
		Tool:      "test",
		Action:    ActionAllow,
	}

	if !d.Timestamp.Equal(now) {
		t.Error("Timestamp mismatch")
	}
}

File: middleware_py/tests/types_test.py
"""Tests for middleware_py.types."""

from middleware_py.middleware_types import (
    Action,
    CallerContext,
    ToolDefinition,
    ToolResult,
    ToolCall,
    Decision,
    RateLimitConfig,
    Condition,
)


class TestAction:
    def test_action_values(self):
        assert Action.ALLOW.value == "allow"
        assert Action.DENY.value == "deny"
        assert Action.FILTER.value == "filter"


class TestCallerContext:
    def test_default_context(self):
        ctx = CallerContext()
        assert ctx.user_id is None
        assert ctx.session_id is None
        assert ctx.trusted is False

    def test_get_method(self):
        ctx = CallerContext(user_id="user1", session_id="sess1", role="admin")
        assert ctx.get("user_id") == "user1"
        assert ctx.get("session_id") == "sess1"
        assert ctx.get("role") == "admin"
        assert ctx.get("nonexistent") is None
        assert ctx.get("nonexistent", "default") == "default"

    def test_metadata_access(self):
        ctx = CallerContext(metadata={"custom_key": "custom_value"})
        assert ctx.get("custom_key") == "custom_value"


class TestToolDefinition:
    def test_minimal_definition(self):
        tool = ToolDefinition(name="test_tool")
        assert tool.name == "test_tool"
        assert tool.description == ""
        assert tool.input_schema == {}

    def test_full_definition(self):
        tool = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )
        assert tool.name == "test_tool"
        assert tool.description == "A test tool"


class TestToolResult:
    def test_successful_result(self):
        result = ToolResult(tool_name="test", success=True, result={"data": "value"})
        assert result.success is True
        assert result.result == {"data": "value"}
        assert result.error is None

    def test_failed_result(self):
        result = ToolResult(tool_name="test", success=False, error="Something went wrong")
        assert result.success is False
        assert result.error == "Something went wrong"


class TestToolCall:
    def test_minimal_tool_call(self):
        call = ToolCall(tool_name="test", args={})
        assert call.tool_name == "test"
        assert call.args == {}

    def test_tool_call_with_context(self):
        ctx = CallerContext(user_id="user1")
        call = ToolCall(tool_name="test", args={"arg1": "val1"}, caller_context=ctx)
        assert call.caller_context.user_id == "user1"


class TestDecision:
    def test_allow_decision(self):
        decision = Decision(action=Action.ALLOW)
        assert decision.action == Action.ALLOW

    def test_deny_with_message(self):
        decision = Decision(
            action=Action.DENY,
            rule_name="test_rule",
            message="Denied by policy",
        )
        assert decision.action == Action.DENY
        assert decision.rule_name == "test_rule"
        assert decision.message == "Denied by policy"


class TestCondition:
    def test_condition_creation(self):
        cond = Condition(field="caller.role", operator="eq", value="admin")
        assert cond.field == "caller.role"
        assert cond.operator == "eq"
        assert cond.value == "admin"


class TestRateLimitConfig:
    def test_rate_limit_config(self):
        config = RateLimitConfig(count=10, window=60)
        assert config.count == 10
        assert config.window == 60